app/src/main/java/app/tivi/ViewModelStore.kt

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package app.tivi
+
+import androidx.compose.runtime.Stable
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.awaitCancellation
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.SharingStarted
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.onCompletion
+import kotlinx.coroutines.flow.stateIn
+import kotlinx.coroutines.isActive
+
+/**
+ * TODO
+ */
+@Stable
+internal class ViewModelStore(
+    coroutineScope: CoroutineScope,
+) {
+    private val viewModelMap = HashMap<Any, ViewModelStoreEntry>()
+
+    // Create a copy of the provided CoroutineScope, but replacing the Job with
+    // a child SupervisorJob. This allows a ViewModel to be cancelled without affecting any
+    // other ViewModels.
+    private val coroutineScope = CoroutineScope(
+        coroutineScope.coroutineContext +
+            SupervisorJob(coroutineScope.coroutineContext[Job])
+    )
+
+    /**
+     * Retrieve or create a ViewModel.
+     */
+    fun <T : Any> viewModelFlow(
+        key: Any,
+        create: (CoroutineScope) -> T
+    ): StateFlow<T> {
```

Comment:
What about changing this API to be `getViewModel(key, create): ViewModel<T>`? What's nice about `StateFlow` is that you can access its `value` synchronously. We don't want to get ViewModel updates from the consumer side (there won't be updates), returning `flow.value` from this function should be enough. Even better, the API is easier to consume and more readable

Reply:
We need the collection of the StateFlow to be in the UI, so that the `WhileSubscribed` works as expected. We could provide a `@Composable` function which returns a `ViewModel` though.

Reply:
Right! That's true. Yeah, probably we need a `@Composable` function to hide this complexity away

Reply:
Done